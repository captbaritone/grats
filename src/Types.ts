import type {
  GraphQLResolveInfo,
  GraphQLScalarLiteralParser,
  GraphQLScalarSerializer,
  GraphQLScalarValueParser,
} from "graphql";

/** @gqlScalar */
export type Float = number;
/** @gqlScalar */
export type Int = number;
/** @gqlScalar */
export type ID = string;

/** @gqlInfo */
export type GqlInfo = GraphQLResolveInfo;

export type GqlScalar<TExternal, TInternal> = {
  /** Serializes an internal value to include in a response. */
  serialize?: GraphQLScalarSerializer<TExternal>;
  /** Parses an externally provided value to use as an input. */
  parseValue?: GraphQLScalarValueParser<TInternal>;
  /** Parses an externally provided literal value to use as an input. */
  parseLiteral?: GraphQLScalarLiteralParser<TInternal>;
};
