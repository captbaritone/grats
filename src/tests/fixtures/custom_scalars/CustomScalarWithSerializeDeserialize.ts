import { GqlScalar } from "../../../Types";

type CustomScalarInputType = string;
type CustomScalarOutputType = number;

/** @gqlScalar Custom */
export const CustomScalar: GqlScalar<
  CustomScalarInputType,
  CustomScalarOutputType
> = {
  serialize(outputValue) {
    // value comes from the server, so it's already in the internal format
    return "10";
  },
  parseValue(value) {
    // value comes from the client, so we need to convert it to the internal format
    return Number(value);
  },
  parseLiteral(ast) {
    if (ast.kind === "StringValue") {
      return Number(ast.value);
    }
    throw new Error(
      "Query error: Can only parse strings to numbers but got a: " + ast.kind,
    );
  },
};
