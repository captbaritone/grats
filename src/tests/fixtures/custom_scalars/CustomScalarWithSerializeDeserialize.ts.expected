-----------------
INPUT
----------------- 
import { GqlScalar } from "../../../Types";

type CustomScalarInputType = string;
type CustomScalarOutputType = number;

/** @gqlScalar Custom */
export const CustomScalar: GqlScalar<
  CustomScalarInputType,
  CustomScalarOutputType
> = {
  serialize(outputValue) {
    // value comes from the server, so it's already in the internal format
    return "10";
  },
  parseValue(value) {
    // value comes from the client, so we need to convert it to the internal format
    return Number(value);
  },
  parseLiteral(ast) {
    if (ast.kind === "StringValue") {
      return Number(ast.value);
    }
    throw new Error(
      "Query error: Can only parse strings to numbers but got a: " + ast.kind,
    );
  },
};

-----------------
OUTPUT
-----------------
-- SDL --
scalar Custom
-- TypeScript --
import { CustomScalar } from "./CustomScalarWithSerializeDeserialize";
import { GraphQLSchema, GraphQLScalarType } from "graphql";
export function getSchema(): GraphQLSchema {
    const CustomType: GraphQLScalarType = new GraphQLScalarType({
        name: "Custom",
        ...CustomScalar
    });
    return new GraphQLSchema({
        types: [CustomType]
    });
}
